app/controllers/species_lists/projects_controller.rb

```diff
@@ -0,0 +1,191 @@
+# frozen_string_literal: true
+
+module SpeciesLists
+  class ProjectsController < ApplicationController
+    before_action :login_required
+    before_action :disable_link_prefetching
+
+    # ----------------------------
+    #  :section: Manage Projects
+    # ----------------------------
+
+    # def manage_projects
+    def edit
+      return unless (@list = find_species_list!)
+
+      if check_permission!(@list)
+        @projects = projects_to_manage
+        @object_states = manage_object_states
+        @project_states = manage_project_states
+      else
+        redirect_to(species_list_path(@list.id))
+      end
+    end
+
+    def update
+      return unless (@list = find_species_list!)
+
+      if check_permission!(@list)
+        @projects = projects_to_manage
+        @object_states = manage_object_states
+        @project_states = manage_project_states
+        commit_and_redirect
+      else
+        redirect_to(species_list_path(@list.id))
+      end
+    end
+
+    def projects_to_manage
+      projects = @user.projects_member
+      if @list.user == @user
+        projects += @list.projects
+        projects.uniq!
+      end
+      projects
+    end
+
+    def manage_object_states
+      {
+        list: params[:objects_list].present?,
+        obs: params[:objects_obs].present?,
+        img: params[:objects_img].present?
+      }
+    end
+
+    def manage_project_states
+      states = {}
+      @projects.each do |proj|
+        states[proj.id] = params["projects_#{proj.id}"].present?
+      end
+      states
```

Comment:
Your choice: (a) implement the following which I like because it's shorter and may be faster, (b) leave as-is.
WARNING: @pellaea dislikes `each_with_object`
```suggestion
      @projects.each_with_object({}) do |proj, states|
        states[proj.id] = params["projects_#{proj.id}"].present?
      end
```

Reply:
Question about `each_with_object`: Does `states` get set to `{}` even if there are no `@projects`?

Reply:
Great question. I hadn't thought about that.
If `@projects` is any Enumerable, including an empty Array, it works fine.
```ruby
irb(main):002:0> @projects = []
=> []
irb(main):003:0> @projects.each_with_object({}) {|proj, states| states[proj.id] = params["projects_#{proj.id}"].present? }
=> {}
```
If @projects is `nil` (or any non-Enumerable) it throws an error
```ruby
irb(main):001:0> @projects.each_with_object({}) {|proj, states| states[proj.id] = params["projects_#{proj.id}"].present? }
(irb):1:in `<main>': undefined method `each_with_object' for nil:NilClass (NoMethodError)
                                                                                                                                               
@projects.each_with_object({}) {|proj, states| states[proj.id] = params["projects_#{proj.id}"].present? }                                      
         ^^^^^^^^^^^^^^^^^                                                                                                                     
```